v2/spiffetls/dial_test.go

```diff
@@ -0,0 +1,301 @@
+			listenMode: spiffetls.TLSServerWithSource(wlAPISourceA),
+		},
+		{
+			name:       "MTLSClientWithSource succeeds",
+			dialMode:   spiffetls.MTLSClientWithSource(tlsconfig.AuthorizeID(serverID), wlAPISourceB),
+			listenMode: spiffetls.MTLSServerWithSource(tlsconfig.AuthorizeID(clientID), wlAPISourceA),
+		},
+		{
+			name:       "MTLSWebClient  succeeds",
+			dialMode:   spiffetls.MTLSWebClientWithSource(webCertPool, wlAPISourceB),
+			listenMode: spiffetls.MTLSWebServerWithSource(tlsconfig.AuthorizeID(clientID), webCert, wlAPISourceA),
+		},
+
+		// *WithSourceOptions Scenario
+		{
+			name:       "TLSClientWithSource succeeds",
+			dialMode:   spiffetls.TLSClientWithSourceOptions(tlsconfig.AuthorizeID(serverID), workloadapi.WithClient(wlAPIClientB)),
+			listenMode: spiffetls.TLSServerWithSource(wlAPISourceA),
+		},
+		{
+			name:       "MTLSClientWithSource succeeds",
+			dialMode:   spiffetls.MTLSClientWithSourceOptions(tlsconfig.AuthorizeID(serverID), workloadapi.WithClient(wlAPIClientB)),
+			listenMode: spiffetls.MTLSServerWithSource(tlsconfig.AuthorizeID(clientID), wlAPISourceA),
+		},
+		{
+			name:       "MTLSWebClient  succeeds",
+			dialMode:   spiffetls.MTLSWebClientWithSourceOptions(webCertPool, workloadapi.WithClient(wlAPIClientB)),
+			listenMode: spiffetls.MTLSWebServerWithSource(tlsconfig.AuthorizeID(clientID), webCert, wlAPISourceA),
+		},
+
+		// *WithRawConfig Scenario
+		{
+			name:       "TLSClientWithSource succeeds",
+			dialMode:   spiffetls.TLSClientWithRawConfig(tlsconfig.AuthorizeID(serverID), bundleSource),
+			listenMode: spiffetls.TLSServerWithSource(wlAPISourceA),
+		},
+		{
+			name:       "MTLSClientWithSource succeeds",
+			dialMode:   spiffetls.MTLSClientWithRawConfig(tlsconfig.AuthorizeID(serverID), svidSource, bundleSource),
+			listenMode: spiffetls.MTLSServerWithSource(tlsconfig.AuthorizeID(clientID), wlAPISourceA),
+		},
+		{
+			name:       "MTLSWebClient  succeeds",
+			dialMode:   spiffetls.MTLSWebClientWithRawConfig(webCertPool, svidSource),
+			listenMode: spiffetls.MTLSWebServerWithSource(tlsconfig.AuthorizeID(clientID), webCert, wlAPISourceA),
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+
+		err := os.Setenv("SPIFFE_ENDPOINT_SOCKET", test.defaultWlAPIAddr)
+		require.NoError(t, err)
+
+		t.Run(test.name, func(t *testing.T) {
+			// Start listening
+			listenCtx, cancelListenCtx := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancelListenCtx()
+
+			var listener net.Listener
+			var listenAddr string
+			listenDoneCh := make(chan struct{})
+			listenDataCh := make(chan string)
+			listenErrCh := make(chan error)
+			if test.listenMode != nil {
+				listener, err = spiffetls.ListenWithMode(listenCtx, "tcp", "localhost:0", test.listenMode)
+				require.NoError(t, err)
+				require.NotNil(t, listener)
+				defer listener.Close()
+
+				listenAddr = listener.Addr().String()
+
+				go func() {
```

Comment:
For test cleanliness sake, we should wait for these goroutines to shut down before returning from the test.